scheduler/cloudformation/migration.go

```diff
@@ -0,0 +1,215 @@
+package cloudformation
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"io"
+
+	"golang.org/x/net/context"
+
+	"github.com/remind101/empire/scheduler"
+	"github.com/remind101/empire/scheduler/ecs"
+)
+
+// This is the environment variable in the application that determines what step
+// of the migration we should transition to. A basic migration flow would look
+// like:
+//
+// 1. `emp set EMPIRE_SCHEDULER_MIGRATION=step1`: CloudFormation stack is
+//    created without any DNS changes.
+// 2. User removes the old CNAME manually in the AWS Console, then sets the
+//    `DNS` parameter in the CloudFormation stack to `true`.
+// 3. `emp set EMPIRE_SCHEDULER_MIGRATION=step2`: The old AWS resources are
+//    removed.
+// 4. `emp unset EMPIRE_SCHEDULER_MIGRATION`: All done.
+const MigrationEnvVar = "EMPIRE_SCHEDULER_MIGRATION"
+
+// ErrMigrating is returned when the application is being migrated.
+var ErrMigrating = errors.New("app is currently being migrated to a CloudFormation stack. Sit tight...")
+
+// This is a scheduler.Scheduler implementation that wraps the newer
+// cloudformation.Scheduler and the older ecs.Scheduler to migrate applications
+// over the the new CloudFormation based scheduler.
+//
+// It uses a sql table to determine what scheduling backend should be used. Apps
+// can be migrated from the ecs scheduler to the cloudformation scheduler by
+// using the Migrate function.
+type MigrationScheduler struct {
+	// The scheduler that we want to migrate to.
+	cloudformation interface {
+		scheduler.Scheduler
+		SubmitWithOptions(context.Context, *scheduler.App, SubmitOptions) error
+	}
+
+	// The scheduler we're migrating from.
+	ecs interface {
+		scheduler.Scheduler
+		RemoveWithOptions(context.Context, string, ecs.RemoveOptions) error
+	}
+
+	db *sql.DB
+}
+
+// NewMigrationScheduler returns a new MigrationSchedeuler instance.
+func NewMigrationScheduler(db *sql.DB, c *Scheduler, e *ecs.Scheduler) *MigrationScheduler {
+	return &MigrationScheduler{
+		db:             db,
+		cloudformation: c,
+		ecs:            e,
+	}
+}
+
+// Backend returns the scheduling backend to use for the given app.
+func (s *MigrationScheduler) Backend(appID string) (scheduler.Scheduler, error) {
+	backend, err := s.backend(appID)
+	if err != nil {
+		return nil, err
+	}
+
+	switch backend {
+	case "ecs":
+		return s.ecs, nil
+	case "cloudformation":
+		return s.cloudformation, nil
+	default:
+		return nil, ErrMigrating
+	}
+}
+
+// backend returns the name of the backend to use for operations.
+func (s *MigrationScheduler) backend(appID string) (backend string, err error) {
+	err = s.db.QueryRow(`SELECT backend FROM scheduler_migration WHERE app_id = $1`, appID).Scan(&backend)
+
+	// For newly created apps.
+	if err == sql.ErrNoRows {
+		backend = "cloudformation"
```

Comment:
Actually this is a named return only because of an existing change I had in here, where the return value was `(string, string, error)`. Generally only used named returns to remove ambiguity in the return values.